airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/message/DestinationMessage.kt

```diff
@@ -7,71 +7,207 @@ package io.airbyte.cdk.message
+            .withRecord(
+                AirbyteRecordMessage()
+                    .withStream(stream.descriptor.name)
+                    .withNamespace(stream.descriptor.namespace)
+                    .withEmittedAt(emittedAtMs)
+                    .withData(data)
+                    .also {
+                        if (meta != null) {
+                            it.meta = meta.asProtocolObject()
+                        }
+                    }
+            )
+}
+
+private fun statusToProtocolMessage(
+    stream: DestinationStream,
+    emittedAtMs: Long,
+    status: AirbyteStreamStatus,
+): AirbyteMessage =
+    AirbyteMessage()
+        .withType(AirbyteMessage.Type.TRACE)
+        .withTrace(
+            AirbyteTraceMessage()
+                .withType(AirbyteTraceMessage.Type.STREAM_STATUS)
+                .withEmittedAt(emittedAtMs.toDouble())
+                .withStreamStatus(
+                    AirbyteStreamStatusTraceMessage()
+                        .withStreamDescriptor(stream.descriptor.asProtocolObject())
+                        .withStatus(status)
+                )
+        )
 
 data class DestinationStreamComplete(
     override val stream: DestinationStream,
-    val emittedAtMs: Long
-) : DestinationRecordMessage()
+    val emittedAtMs: Long,
+) : DestinationStreamAffinedMessage {
+    override fun asProtocolMessage(): AirbyteMessage =
+        statusToProtocolMessage(stream, emittedAtMs, AirbyteStreamStatus.COMPLETE)
+}
+
+data class DestinationStreamIncomplete(
+    override val stream: DestinationStream,
+    val emittedAtMs: Long,
+) : DestinationStreamAffinedMessage {
+    override fun asProtocolMessage(): AirbyteMessage =
+        statusToProtocolMessage(stream, emittedAtMs, AirbyteStreamStatus.INCOMPLETE)
+}
 
 /** State. */
-sealed class CheckpointMessage : DestinationMessage() {
+sealed interface CheckpointMessage : DestinationMessage {
     data class Stats(val recordCount: Long)
     data class StreamCheckpoint(
         val stream: DestinationStream,
         val state: JsonNode,
     )
 
-    abstract val sourceStats: Stats
-    abstract val destinationStats: Stats?
+    val sourceStats: Stats
+    val destinationStats: Stats?
 
-    abstract fun withDestinationStats(stats: Stats): CheckpointMessage
+    fun withDestinationStats(stats: Stats): CheckpointMessage
 }
 
+private fun streamCheckpointToProtocolObject(
+    streamCheckpoint: CheckpointMessage.StreamCheckpoint
+): AirbyteStreamState =
+    AirbyteStreamState()
+        .withStreamDescriptor(streamCheckpoint.stream.descriptor.asProtocolObject())
+        .withStreamState(streamCheckpoint.state)
+
 data class StreamCheckpoint(
-    val streamCheckpoint: StreamCheckpoint,
-    override val sourceStats: Stats,
-    override val destinationStats: Stats? = null
-) : CheckpointMessage() {
-    override fun withDestinationStats(stats: Stats) =
+    val streamCheckpoint: CheckpointMessage.StreamCheckpoint,
```

Comment:
I don't really understand how this was working previously. Explicitly use the child class here to avoid a recursive constructor

(I can't think of a better name to avoid this clash :shrug: )

Reply:
Yeah that must have happened during the renaming. Weirdly seems to resolve correctly though?

Either Change the StreamCheckpoint/GlobalCheckpoint child classes to StreamCheckpointMessage and GlobalCheckpointMessage or change the StreamCheckpoint inner class to Checkpoint?